Source/santa-driver/SantaPrefixTree.h

```diff
@@ -0,0 +1,82 @@
+/// Copyright 2018 Google Inc. All rights reserved.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///    http://www.apache.org/licenses/LICENSE-2.0
+///
+///    Unless required by applicable law or agreed to in writing, software
+///    distributed under the License is distributed on an "AS IS" BASIS,
+///    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+///    See the License for the specific language governing permissions and
+///    limitations under the License.
+
+#ifndef SANTA__SANTA_DRIVER__SANTAPREFIXTREE_H
+#define SANTA__SANTA_DRIVER__SANTAPREFIXTREE_H
+
+#include <IOKit/IOLib.h>
+#include <libkern/c++/OSObject.h>
+#include <libkern/locks.h>
+
+///
+///  SantaPrefixTree is a simple prefix tree implementation.
+///  Adding and checking prefixes are thread safe.
+///
+class SantaPrefixTree : public OSObject {
+  OSDeclareDefaultStructors(SantaPrefixTree);
+
+ public:
+  bool init() override;
+  void free() override;
+
+  void AddPrefix(const char *);
+  bool HasPrefix(const char *);
+  // TODO(bur): Add RemoveAll(). This will allow santad to reset / add prefixes while running.
+
+ private:
+  ///
+  ///  SantaPrefixNode is a wrapper class that represents one byte.
+  ///  1 node can represent a whole ASCII character.
+  ///  For example a pointer to the 'A' node will be stored at children[0x41].
+  ///  It takes 1-4 nodes to represent a UTF8 encoded Unicode character.
+  ///
+  ///  The path for "/🤘" would look like this:
+  ///      children[0x2f] -> children[0xf0] -> children[0x9f] -> children[0xa4] -> children[0x98]
+  ///
+  ///  The path for "/dev" is:
+  ///      children[0x2f] -> children[0x64] -> children[0x65] -> children[0x76]
+  ///
+  ///  Lookups of children are O(1).
+  ///
+  ///  Having the nodes represented by a smaller width, such as a nibble (1/2 byte), would
+  ///  drastically decrease the memory footprint but would double required dereferences.
+  ///
+  class SantaPrefixNode {
+   public:
+    bool isPrefix;
+    void *children[256];
```

Comment:
It's probably not worth the 9 bytes that you'd save per node, but children[0] will never be used, so you could declare an array of length 255 and adjust the indexing.

Reply:
Ack